components/esmterrain/storage.hpp

```diff
@@ -137,6 +139,103 @@ namespace ESMTerrain
         bool mAutoUseSpecularMaps;
 
         Terrain::LayerInfo getLayerInfo(const std::string& texture);
+
+    protected:
+
+        inline void fixNormal (osg::Vec3f& normal, int cellX, int cellY, int col, int row, LandCache& cache)
```

Comment:
Methods are inline by default if defined inside a class definition. `inline` keyword is not required here.

Reply:
Thanks, I was really confused of this, and this is the part where I think I need most advice. I would actually rather not change anything at /components/esmterrain/storage.cpp or /components/esmterrain/storage.hpp. 

I only changed it because in [terrainstorage.hpp](https://github.com/OpenMW/openmw/blob/685bcca3c23be13434112ebf202abdbe98cee5e4/apps/opencs/view/render/terrainstorage.hpp#L33) I have fillVertexBuffers that _overrides_ one in storage.hpp. Functions fixNormal etc. are used by fillVertexBuffers. Can you think of a way to override void fillVertexBuffers without touching components/esmterrain/storage.hpp?

Reply:
As I understand the only difference is [this](https://github.com/OpenMW/openmw/blob/bb679ca702bc4fbccdd66b328c3af2db89d24743/apps/opencs/view/render/terrainstorage.cpp#L156) line and [these](https://github.com/OpenMW/openmw/blob/bb679ca702bc4fbccdd66b328c3af2db89d24743/apps/opencs/view/render/terrainstorage.cpp#L192-L215) lines. Simple solution would be to add a new virtual methods to `ESMTerrain::Storage` that will do nothing by default and will have overrides in `CSVRender::TerrainStorage`. At least this allows to avoid logic duplication.

```diff
diff --git a/apps/opencs/view/render/terrainstorage.hpp b/apps/opencs/view/render/terrainstorage.hpp
index 5f808d843..12f76a50a 100644
--- a/apps/opencs/view/render/terrainstorage.hpp
+++ b/apps/opencs/view/render/terrainstorage.hpp
@@ -29,14 +29,9 @@ namespace CSVRender
 
         virtual osg::ref_ptr<const ESMTerrain::LandObject> getLand (int cellX, int cellY) override;
         virtual const ESM::LandTexture* getLandTexture(int index, short plugin) override;
-
-        /// Draws temporarily altered land (transient change support)
-        void fillVertexBuffers (int lodLevel, float size, const osg::Vec2f& center,
-                        osg::ref_ptr<osg::Vec3Array> positions,
-                        osg::ref_ptr<osg::Vec3Array> normals,
-                        osg::ref_ptr<osg::Vec4ubArray> colours) override;
-
         virtual void getBounds(float& minX, float& maxX, float& minY, float& maxY) override;
+        void adjustColor(int col, int row, const ESM::Land::LandData *heightData, osg::Vec4ub& color) const override;
+        float getAlteredHeight(int col, int row) const override;
     };
 
 }
diff --git a/components/esmterrain/storage.cpp b/components/esmterrain/storage.cpp
index 151f0ff91..bfdca27d3 100644
--- a/components/esmterrain/storage.cpp
+++ b/components/esmterrain/storage.cpp
@@ -172,7 +172,7 @@ namespace ESMTerrain
                         (*positions)[static_cast<unsigned int>(vertX*numVerts + vertY)]
                             = osg::Vec3f((vertX / float(numVerts - 1) - 0.5f) * size * Constants::CellSizeInUnits,
                                          (vertY / float(numVerts - 1) - 0.5f) * size * Constants::CellSizeInUnits,
-                                         height);
+                                         height + getAlteredHeight(col, row));
 
                         if (normalData)
                         {
@@ -208,6 +208,8 @@ namespace ESMTerrain
                             color.b() = 255;
                         }
 
+                        adjustColor(col, row, heightData, color);
+
                         // Unlike normals, colors mostly connect seamlessly between cells, but not always...
                         if (col == ESM::Land::LAND_SIZE-1 || row == ESM::Land::LAND_SIZE-1)
                             fixColour(color, cellX, cellY, col, row, cache);
@@ -501,4 +503,13 @@ namespace ESMTerrain
         return ESM::Land::LAND_TEXTURE_SIZE*chunkSize;
     }
 
+    void Storage::adjustColor(int col, int row, const ESM::Land::LandData *heightData, osg::Vec4ub& color) const
+    {
+    }
+
+    float Storage::getAlteredHeight(int col, int row) const
+    {
+        return 0;
+    }
+
 }
```

Reply:
So just that I understand you correctly: You mean that I'd define something like `void myFunction()` at `ESMTerrain::Storage`, and *always* call in it in `ESMTerrain::Storage::fillVertexBuffers`, yet it would do nothing in the base class? Then I'd define `CSVRender::TerrainStorage::myFunction` that overrides the function? If it works, it avoids logic duplication and is a whole lot cleaner.

Reply:
Fixed this in commit f7e0064 .